src/asymmetric_encryption.cpp

```diff
@@ -0,0 +1,242 @@
+                break;
+            }
+            default: throw MoCOCrWException("Unsupported encryption mode");
+        }
+
+        /* Validate message size */
+        if (static_cast<int>(message.toByteArray().size()) > maxSize) {
+            throw MoCOCrWException("Message too long for RSA key size");
+        }
+        _CRYPTO_malloc_init();
+        _EVP_PKEY_encrypt_init(keyCtx.get());
+        _EVP_PKEY_CTX_set_rsa_padding(keyCtx.get(), static_cast<int>(paddingMode));
+
+        if (RSAPaddingMode::OAEP == paddingMode) {
+            const auto oaepPaddingMode = static_cast<const OAEPPadding&>(pad);
+
+            _EVP_PKEY_CTX_set_rsa_oaep_md(keyCtx.get(),
+                                    _getMDPtrFromDigestType(oaepPaddingMode.getHashingFunction()));
+
+            _EVP_PKEY_CTX_set_rsa_mgf1_md(keyCtx.get(),
+                                    _getMDPtrFromDigestType(oaepPaddingMode.getMaskingFunction()));
+
+            if(oaepPaddingMode.getLabel().size() > 0){
+
+                /* Make a copy of the label, since the context takes ownership of it when calling
+                 * '_EVP_PKEY_CTX_set_rsa_oaep_label()' function*/
+                label_copy.reset(static_cast<uint8_t*>(
+                                            _OPENSSL_malloc(oaepPaddingMode.getLabel().size())));
+                memcpy(label_copy.get(),
+                                &oaepPaddingMode.getLabel()[0], oaepPaddingMode.getLabel().size());
+
+                _EVP_PKEY_CTX_set_rsa_oaep_label(keyCtx.get(),
+                                             static_cast<unsigned char*>(label_copy.get()),
+                                             static_cast<int>(oaepPaddingMode.getLabel().size()));
+
+                /* Release ownership from the unique_ptr since the function above takes ownership of
+                 * the label pointer unless it throws an exception. Unique_ptr should only keep
+                 * ownership if an exception is thrown by _EVP_PKEY_CTX_set_rsa_oaep_label.*/
+                label_copy.release();
+            }
+        }
+
+        /* First call to determine the buffer length */
+        _EVP_PKEY_encrypt(keyCtx.get(),
+                          nullptr,
+                          &encryptedMessageLen,
+                          reinterpret_cast<const unsigned char *>(&message.toByteArray()[0]),
+                          message.toByteArray().size());
+
+        /* Allocate memory for the buffer, based on the size returned by _EVP_PKEY_encrypt */
+        encryptedMessage.reset(static_cast<unsigned char*>(
+                                        _OPENSSL_malloc(static_cast<int>(encryptedMessageLen))));
+
+        /* Second call to perform the actual encryption */
+        _EVP_PKEY_encrypt(keyCtx.get(),
+                          encryptedMessage.get(),
+                          &encryptedMessageLen,
+                          reinterpret_cast<const unsigned char *>(&message.toByteArray()[0]),
```

Comment:
message.data()

Reply:
I miss-understood your comment original. Changed to use data().